😹😹 Improve tokenization

louib opened the pull request:
Hi again @mrowa44 @GrenderG :wave:

This PR improves the handling of emojis inside words boundaries,
or just emojis not separated by spaces, by using `sed` for the tokenization.

I also added added a `bats` test file, to prevent regressions of the
current behaviour :smile_cat: 
```
$ bats emojify_tests.bats 
 ✓ handles an input without a single emoji 😿
 ✓ handles an input with a single emoji 😹
 ✓ handles an input with a lot of emojis 😻
 ✓ handles emojis with underscores and numbers 💯
 ✓ handles emojis with + and - 👍
 ✓ handles punctuations just after aliases
 ✓ handles multiple spaces after an emoji
 ✓ handles the examples from the readme 😉

8 tests, 0 failures
```

I think this PR will close both https://github.com/mrowa44/emojify/issues/24 and https://github.com/mrowa44/emojify/issues/37, as suggested by the new `bats` test cases.

Thank you!

louib commented:
@GrenderG should I add a small section on how to run the tests in the README??

GrenderG commented:
> @GrenderG should I add a small section on how to run the tests in the README??

Sure!

louib commented:
@GrenderG there you go, added a small section in the README. Thanks for the review!

mrowa44 commented:
Hey @louib! Thanks a lot for your PR. I see you've used `sed -r`, this doesn't really work on my machine:

```sh
bash-5.0$ bats emojify_tests.bats
 ✗ handles an input without a single emoji 😿
   (in test file emojify_tests.bats, line 5)
     `[ "$result" = "no emoji :(" ]' failed
   sed: illegal option -- r
   usage: sed script [-Ealn] [-i extension] [file ...]
          sed [-Ealn] [-i extension] [-e script] ... [-f script_file] ... [file ...]
 ✗ handles an input with a single emoji 😹
   (in test file emojify_tests.bats, line 10)
     `[ "$result" = "an emoji 😁" ]' failed
   sed: illegal option -- r
   usage: sed script [-Ealn] [-i extension] [file ...]
          sed [-Ealn] [-i extension] [-e script] ... [-f script_file] ... [file ...]
 ✗ handles an input with a lot of emojis 😻
   (in test file emojify_tests.bats, line 15)
     `[ "$result" = "emojis 😁😁 🎉yay🍾" ]' failed
   sed: illegal option -- r
   usage: sed script [-Ealn] [-i extension] [file ...]
          sed [-Ealn] [-i extension] [-e script] ... [-f script_file] ... [file ...]
 ✗ handles emojis with underscores and numbers 💯
   (in test file emojify_tests.bats, line 20)
     `[ "$result" = "this is perfect 💯 🥇" ]' failed
   sed: illegal option -- r
   usage: sed script [-Ealn] [-i extension] [file ...]
          sed [-Ealn] [-i extension] [-e script] ... [-f script_file] ... [file ...]
 ✗ handles emojis with + and - 👍
   (in test file emojify_tests.bats, line 25)
     `[ "$result" = "great 👍👍👎" ]' failed
   sed: illegal option -- r
   usage: sed script [-Ealn] [-i extension] [file ...]
          sed [-Ealn] [-i extension] [-e script] ... [-f script_file] ... [file ...]
 ✗ handles punctuations just after aliases
   (in test file emojify_tests.bats, line 30)
     `[ "$result" = "Enter the ✈️!" ]' failed
   sed: illegal option -- r
   usage: sed script [-Ealn] [-i extension] [file ...]
          sed [-Ealn] [-i extension] [-e script] ... [-f script_file] ... [file ...]
 ✗ handles multiple spaces after an emoji
   (in test file emojify_tests.bats, line 35)
     `[ "$result" = "✨   Three spaces" ]' failed
   sed: illegal option -- r
   usage: sed script [-Ealn] [-i extension] [file ...]
          sed [-Ealn] [-i extension] [-e script] ... [-f script_file] ... [file ...]
 ✗ handles the examples from the readme 😉
   (in test file emojify_tests.bats, line 44)
     `[ "$result" = "Hey, I just 🙋 you, and this is 😱 , but here's my 📲 , so 📞 me, maybe?" ]' failed
   sed: illegal option -- r
   usage: sed script [-Ealn] [-i extension] [file ...]
          sed [-Ealn] [-i extension] [-e script] ... [-f script_file] ... [file ...]

8 tests, 8 failures
```

Do you know what may be the reason? I found this site: http://pubs.opengroup.org/onlinepubs/9699919799/utilities/sed.html from 2018 it seems and there's no `-r` option there.

@GrenderG can you run the tests locally?

louib commented:
@mrowa44 thanks for testing!

I removed the `-r` flag that was forcing the extended regex syntax, and changed to the standard `sed` regex syntax instead. Could you give it another try please :pray:

mrowa44 commented:
Yes, now the sed command isn't failing anymore but the tests are :slightly_smiling_face:

```bash
bash-5.0$ bats emojify_tests.bats
 ✓ handles an input without a single emoji 😿
 ✗ handles an input with a single emoji 😹
   (in test file emojify_tests.bats, line 10)
     `[ "$result" = "an emoji 😁" ]' failed
 ✗ handles an input with a lot of emojis 😻
   (in test file emojify_tests.bats, line 15)
     `[ "$result" = "emojis 😁😁 🎉yay🍾" ]' failed
 ✗ handles emojis with underscores and numbers 💯
   (in test file emojify_tests.bats, line 20)
     `[ "$result" = "this is perfect 💯 🥇" ]' failed
 ✗ handles emojis with + and - 👍
   (in test file emojify_tests.bats, line 25)
     `[ "$result" = "great 👍👍👎" ]' failed
 ✗ handles punctuations just after aliases
   (in test file emojify_tests.bats, line 30)
     `[ "$result" = "Enter the ✈️!" ]' failed
 ✗ handles multiple spaces after an emoji
   (in test file emojify_tests.bats, line 35)
     `[ "$result" = "✨   Three spaces" ]' failed
 ✗ handles the examples from the readme 😉
   (in test file emojify_tests.bats, line 44)
     `[ "$result" = "Hey, I just 🙋 you, and this is 😱 , but here's my 📲 , so 📞 me, maybe?" ]' failed

8 tests, 7 failures
```

If I run emojify command it seems to only work with emoji alone (`./emojify ":grin:"`, not `./emojify ":grin: test"`)
```bash
bash-5.0$ ./emojify ":sparkles:   Three spaces"
:sparkles:   Three spaces
bash-5.0$ ./emojify "an emoji :grin:"
an emoji :grin:
bash-5.0$ ./emojify "emojis :grin::grin: :tada:yay:champagne:"
emojis :grin::grin: :tada:yay:champagne:
bash-5.0$ ./emojify ":grin:"
😁
bash-5.0$ ./emojify ":grin: test"
:grin: test
```

louib commented:
@mrowa44 :crying_cat_face: I'll have to test on MacOS, looks like the sed command is quite different indeed. I'm not sure there's a way that's going to work for both OS, so I added a branching on the OS already.

GrenderG commented:
Works perfectly for me on GNU/Linux.

louib commented:
@mrowa44 The MacOS version of the tokenization should now be working. Can you give it one last try please :pray:

mrowa44 commented:
@GrenderG I'm not sure if we want to depend on Perl, what do you think?

@louib it's difficult to review the changes you make and pull them locally if you force push. It's good to have a history of changes, we can always squash when merging.

louib commented:
@mrowa44 you're right, sorry for the early squash

GrenderG commented:
> @GrenderG I'm not sure if we want to depend on Perl, what do you think?

I'd rather not depend on Perl, but the changes made are very useful... Does macOS come with `perl` shipped by default, @mrowa44? 

Another solution could be using `gnu-sed` for macOS. We can check if the user has `gnu-sed` installed and use it, and if not, just do it using our old way. We'd need to write in the README file that for better support on macOS users can optionally install `gnu-sed`, then.

louib commented:
@GrenderG @mrowa44 There also the possibility to do the tokenization in pure bash. Would that be an acceptable solution?

GrenderG commented:
> @GrenderG @mrowa44 There's also the possibility to do the tokenization in pure bash. Would that be an acceptable solution?

Yeah! I think so.

louib commented:
I pushed a new commit with the pure bash tokenization version, let me know what you think!

I added a new unit test for a case that was not handled by the `sed` tokenization, and the tests are still passing on :penguin:, and let's :crossed_fingers: on :green_apple: too!
```
$ bats emojify_tests.bats 
 ✓ handles an input without a single emoji 😿
 ✓ handles an input with a single emoji 😹
 ✓ handles an input with a lot of emojis 😻
 ✓ handles emojis with underscores and numbers 💯
 ✓ handles emojis with + and - 👍
 ✓ handles right-hand side emojis 👉
 ✓ handles punctuations just after aliases
 ✓ handles multiple spaces after an emoji
 ✓ handles the examples from the readme 😉

9 tests, 0 failures
```

I also emojified the function names(now `emojify_line` and
`emojify_token`) :exclamation: :exclamation: 

If there's any other test you can think of please let me know!

mrowa44 commented:
@GrenderG this seems to be working fine for me now:

```sh
bash-5.0$ bats emojify_tests.bats
 ✓ handles an input without a single emoji 😿
 ✓ handles an input with a single emoji 😹
 ✓ handles an input with a lot of emojis 😻
 ✓ handles emojis with underscores and numbers 💯
 ✓ handles emojis with + and - 👍
 ✓ handles right-hand side emojis 👉
 ✓ handles punctuations just after aliases
 ✓ handles multiple spaces after an emoji
 ✓ handles the examples from the readme 😉

9 tests, 0 failures
bash-5.0$ ./emojify ":grin: test"
😁 test
```

GrenderG commented:
Yes, it works for me too @mrowa44. But with the current implementation issue #37 is not being solved. Better than nothing, though!

louib commented:
@GrenderG 
>  But with the current implementation issue #37 is not being solved. 

Really? I thought the `handles multiple spaces after an emoji` unit test was testing just that :thinking:. Are you using the example from the issue?

GrenderG commented:
Yeah I'm using that example, @louib.

mrowa44 commented:
Thanks a lot @louib for this! I released it as 2.1.0 https://github.com/mrowa44/emojify/releases/tag/2.1.0

And thank you @GrenderG for helping me maintain this project!

louib commented:
@mrowa44 my pleasure, really :blush: 
thanks to both of you for taking the time to review!
